(161) Residents can book a repair appointment

leeky opened the pull request:
Allow a resident to choose from a list of available appointments. 

NB: Whilst the tests pass, the API doesn't yet provide the real endpoints, so this cannot be run 'live'. Also, the stubbed endpoints are subject to change.

dgmstuart committed:
(feature) dummy appointments page

Works just enough to get us through to the confirmation page.

dgmstuart committed:
(spec) Residents can book appointments

leeky committed:
(feature) HackneyApi can request available appointments for a work order

leeky committed:
(feature) Appointments returned from API can be presented

After requesting a list of appointments from the API, wrap each response
in a `AppointmentPresenter` which provides methods for showing the
available slots in a more human-readable format.

leeky committed:
(feature) Allow user to pick from a list of available appointments

The routes in the Rails app for appointments now reference the
`repair_request_reference` rather than the `work_order_reference`. This
is because we can always use the API to fetch the latter with the
former, but not the other way around.

I've reused the `Repair` class to extract the references from the
response from the API.

leeky committed:
(feature) Allow user to book their chosen appointment

leeky committed:
(feature) Store chosen appointment so it can be shown in summary

Store the chosen appointment into the `SelectedAnswerStore` and display
it to the user on the confirmation page.

leeky committed:
(feature) When it's something o'clock, omit :00

e.g if it's "13:00", display "1pm" instead of "1:00pm"

dgmstuart commented on app/presenters/appointment_presenter.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def description
+    "#{day_of_week} #{begin_time}-#{end_time} (#{day} #{month})"
```
Nicely broken out 👍

dgmstuart commented on app/controllers/appointments_controller.rb:
```diff
@@ -1,11 +1,41 @@
+    params.require(:appointment_form).permit(:appointment_id)
+  end
+
+  def work_order_reference
```
The original idea was that `Appointments` would be a facade exposing `form` and `slots_for_collection` and providing a place for these methods to live so that we don't have to add them in the controller.

This approach works for now but would it be more flexible to move these methods into an object?

leeky replied:
I'm going to leave this in the controller for now I think. If it gets any more complex, I'll revisit and extract this into smaller objects.

dgmstuart commented on app/models/appointment_form.rb:
```diff
@@ -8,4 +8,18 @@ class AppointmentForm
+  private
+
+  def split_date
+    @split_date ||= @appointment_id.to_s.split('|')
```
Is this because the ID will be of form `firstdate|lastdate`? Does this need to be shown in a test somewhere?

leeky replied:
Yeah, this exists because appointments don't have IDs - in fact, we still have no idea whether start/end time is enough to book with - I will add some tests around this.

dgmstuart commented on app/presenters/appointment_presenter.rb:
```diff
@@ -51,10 +51,20 @@ def day
+    friendly_time(end_date)
+  end
+
+  def friendly_time(time)
```
Could this be extracted to an initializer? e.g. https://robots.thoughtbot.com/custom-formats-for-datetime

leeky replied:
I thought about this, but I've read the GOVUK guidelines that suggest using "midday" in place of "12pm", which seems like a good idea - and this wouldn't work with Rails' standard datetime formatting.

I'll add a card to add the 'midday' special case in, and leave this method as-is for now(?)

dgmstuart commented on spec/presenters/appointment_presenter_spec.rb:
```diff
@@ -49,7 +59,7 @@
       )
 
-      expect(presenter.time).to eql '12:00pm and 4:45pm'
+      expect(presenter.time).to eql '12pm and 4:45pm'
```
You've handled this time display nicely 👍

dgmstuart reviewed (dismissed):
Can't test from over here, but looks good - good job 👍

leeky committed:
(spec) Ensure that `appointment_id` is parsed correctly

Add tests around extracting the `begin_date` and `end_date` parameters
from the `appointment_id` field. I've added some additional sanity
checking that throws an error if the field isn't formatted correctly
(which would only happen in the case of malicious behaviour).

leeky commented:
@dgmstuart Should be ready for another look.

techbelly commented:
This looks ok to me. One concern which isn't enough for me to reject it.

If I've understood it correctly, I'm a tiny bit nervous that we present the user a list of appointments, and they could pass us back a start time and end time that might not be in that list. That's nice and stateless, which is good. But it's open to user tampering.

There are a couple of simple ways around this - we could cryptographically sign the ids, I think there's some support in rails for that. We could also store the appointment list in the session and index into it.

I'm not sure I'm worried about this in this case. Up to you if you want to take that further.

techbelly reviewed (approved):
Added a comment on general approach to ids. Code looks good to me, and I understand your responses to duncan's questions.
